types/react-navigation/index.d.ts

```diff
@@ -86,7 +87,7 @@ export type NavigationLeafRoute<Params> = {
    * Params passed to this route when navigating to it,
    * e.g. `{ car_id: 123 }` in a route that displays a car.
    */
-  params?: Params,
+  params: Params,
```

Comment:
Hm, this seems wrong to me. If we're typing out something, we probably want it to be as accurate as possible, even if that isn't quite as convenient sometimes. Here, you're suggesting trading off accuracy for convenience it seems? I think that this library/typing should be generic and not make assumptions about its use - just stick as close to what the actual underlying types are at runtime as possible.

I do understand it's a bit inconvenient, but why not just cast each time you need to access `params`? For example, `const params: ThisComponentsNavigationParamsInterface = this.props.navigation.state.params`?

Lastly, I think the biggest point is that this can easily result in trying to access an undefined object in many cases. Consider the following example.

```
interface ThisComponentsNavigationParamsInterface {
    something?: string,
}

...

render() {
    if (this.props.navigation.state.params.something) {
       return ( <Text>...</Text> );
    } else {
       return ( <View>...</View> );
    }
}

...

<From some other component>
`this.props.navigation.navigate(ThisComponentRouteName);
```

This would result in a runtime exception when accessing `state.params.something`, since `params` is actually totally empty, NOT an empty object. So in this case you'd have to check for `state.params && state.params.something`, which then is as inconvenient as you were saying before, just in the other direction.

You could argue that this is still somewhat the user's fault, since maybe they should have allowed `something` to also be `null`, and then passed it in explicitly in the `navigate` call as `null`, but this seems like we're building up more gotchas than solving problems for the general user, especially if they're new.

Thoughts?